lib/chef/node/mixin/state_tracking_array.rb

```diff
@@ -0,0 +1,41 @@
+#--
+# Copyright:: Copyright (c) Chef Software Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require_relative "immutablize_array"
+require_relative "state_tracking"
+
+class Chef
```

Comment:
Why this approach and not just an adjustment to `Chef::Node::AttrArray` change? Either way, I think it would be helpful to see this illustrated in tests, and definitely would want tests to cover this change.

Reply:
What do you mean by "why this approach"?

If you mean "Why did you added a distinct mixin module?", the answer is quite trivial, I naively replicated what was done; the tracking feature was already prepended on top of the `AttrArray` class, I think using a different approach for the Array specificities would be confusing - especially as there are dedicated mixin for most features (mash, immutabilize); also some of the method are explicitly "hardcoded in `AttrArray`, while most of them are auto-defined (like below); so it would introduce some inconsistency I guess.
=> IMHO it is not that bad to have a mixin, if you think I should just adapt `AttrArray`, let me know I'll do it

Reply:
About tests, do you have some hint on what kind of test you would like to see?
The thing is `AttrArray` is currently not directly tested, as this is kind of the low level implem of the node collection; try to  test the integration with the state tracking feature (also untested) could be great, but not trivial.
My initial idea was to rely solely on the `NodeAttributes` test, I understand it might not be enough.

Reply:
I'm thinking a unit test of a class that uses the changes and then write out a couple of use cases that inspired this... but I see that `StateTracking` isn't validated directly today... at most it looks like `unit/node/vivid_mash_spec.rb` calls out the AttrArray.

Reply:
My main concern was around potential changes in DeepMergeCache behaviour, but after checking it looks like I added tests around AttrArray in https://github.com/chef/chef/pull/13844/files#diff-dffca7cd72dfb684256c4eb165d66cfed8cd089643946b7fd8668a412a11f7d4 so if those pass then I think this should be okay.

Reply:
OK so @tpowell-progress or @dafyddcrosby is there anything I should do to improve this PR?

Reply:
@Annih Looking around at tests, it looks like everything but OpenSUSE LEAP is passing. That's a known issue, but if you want to green, rebasing on `chef/chef:main` will resolve that one. Otherwise, we'll double check that we're ready to merge at tomorrow's community meeting (unless @dafyddcrosby checks in before then)

Reply:
OK @tpowell-progress , I rebased and all checks are green!